# Daily assignments
These are the questions selected for marking. Reminder, answers should be <= 1 paragraph in length. 
Answers should be sent to gregory.owens@alumni.ubc.ca

2. Why is piping directly between programs faster than writing each consecutive output to the disk? Explain using information about computer hardware.

3. Try different filtering options for the GBS data  (see http://prinseq.sourceforge.net/manual.html for options) and plot QC graphs. Discuss which options you would choose to implement if this was your data and why.

4. Is an alignment with a higher percent of mapped reads always better than one with a lower percent? Why or why not?

5. Quantify the assembly metrics for your first assembly that you ran without any options (sa_assembly21). Pick two or three different sets of parameters to run. Compare the resulting assemblies with one another and discuss which ones seemed to have improved the assembly and why that might be. 

### Questions 2-5 are due on Monday, July 10th at 10 am. 

6. What expression measure would you use to compare gene expression between different genes and why? Is it appropriate to compare the raw expression counts? Can you get more appropriate data from RSEM?

7. You're trying to create a very stringent set of SNPs. Based on the site information GATK produces, what filters would you use? Include the actual GATK abbreviations.

8. For a site that is invariant in both populations (i.e. a locus with no variation), what is Fst?

9. If you have a dataset of 100 samples and 100,000 SNPs, what is the maximum number of PC axes?

10. What does it mean when something has 50% bootstrap support? What are two possible reasons that a node may have low support? Include one biological and one methodological reason.

### Questions 6-10 are due on Monday July 17th at 10 am. 


